pkg/client/apiutil/restmapper.go

```diff
@@ -230,35 +259,36 @@ func (m *mapper) addKnownGroupAndReload(groupName string, versions ...string) er
 	}
 
 	m.mapper = restmapper.NewDiscoveryRESTMapper(updatedGroupResources)
-	return nil
 }
 
-// findAPIGroupByNameLocked returns API group by its name.
-func (m *mapper) findAPIGroupByName(groupName string) (*metav1.APIGroup, error) {
-	// Looking in the cache first.
+// findAPIGroupByNameAndMaybeAggregatedDiscovery tries to finds the passed apiGroup.
+// If the server supports aggregated discovery, it will always perform that.
+func (m *mapper) findAPIGroupByNameAndMaybeAggregatedDiscovery(groupName string) (_ *metav1.APIGroup, didAggregatedDiscovery bool, _ error) {
+	// Looking in the cache first
 	{
 		m.mu.RLock()
 		group, ok := m.apiGroups[groupName]
 		m.mu.RUnlock()
 		if ok {
-			return group, nil
+			return group, false, nil
 		}
 	}
 
 	// Update the cache if nothing was found.
-	apiGroups, err := m.client.ServerGroups()
+	apiGroups, maybeResources, _, err := m.client.GroupsAndMaybeResources()
 	if err != nil {
-		return nil, fmt.Errorf("failed to get server groups: %w", err)
+		return nil, false, fmt.Errorf("failed to get server groups: %w", err)
 	}
 	if len(apiGroups.Groups) == 0 {
-		return nil, fmt.Errorf("received an empty API groups list")
+		return nil, false, fmt.Errorf("received an empty API groups list")
 	}
 
 	m.mu.Lock()
-	for i := range apiGroups.Groups {
-		group := &apiGroups.Groups[i]
-		m.apiGroups[group.Name] = group
+	m.initialDiscoveryDone.Store(true)
+	if len(maybeResources) > 0 {
+		didAggregatedDiscovery = true
 	}
+	m.addGroupVersionResourcesToCache(maybeResources)
```

Comment:
Follow-up to: https://github.com/kubernetes-sigs/controller-runtime/pull/2901#discussion_r1706902772

We can also only call addGroupVersionResourcesToCache if it is actually necessary, e.g.:
```go
	m.initialDiscoveryDone.Store(true)
	if len(maybeResources) > 0 {
		didAggregatedDiscovery = true
		m.mu.Lock()
		m.addGroupVersionResourcesToCache(maybeResources)
		m.mu.Unlock()
	}
```

Reply:
Done